### PR TITLE
Improve management of relevant terms in TheoryStrings

### PR DESCRIPTION
**Pull request by ajreynol**
This makes the computation of relevant terms cached per full effort check.
It also corrects an issue where terms that are not relevant for model building were not considered relevant.

Finally, it makes it so that string reductions are only applied for active *and relevant* terms. This change currently has very little impact since we eagerly add lemmas for all preregistered terms (i.e. preregistered terms and terms in assertions are roughly equivalent).

I did an SMT-LIB run with this change and there was no difference in solved benchmarks or significant differences in time.

This is in preparation for further testing for making our preregistration lemmas done lazier.

### CODE REVIEW

File: src/theory/strings/term_registry.h [hunk 1]
```diff
@@ -232,8 +232,19 @@ class TermRegistry : protected EnvObj
...
+   */
+  void computeRelevantTermSet();
+  /**
+   * Get the relevant term set, returns the set computed by the above function.
+   *
+   * The relevant term set is all terms that belong to theory of strings that
+   * appear in the current set of assertions, or are marked as shared terms
+   * for the theory of strings.
+   */
+  const std::set<Node>& getRelevantTermSet() const;
```
Comment (yoni206) [inline src/theory/strings/term_registry.h:247]: Is this separation between `compute` and `get` standard in our codebase? I am a bit worried of someone calling `get` while forgetting to call `compute` before. Can we protect ourselves against something like this using some assertion inside `get`?

File: src/theory/strings/base_solver.cpp [hunk 2]
```diff
@@ -215,8 +215,8 @@ void BaseSolver::checkInit()
               // assuming that the reduction of f(a) depends on itself.
             }
             // this node is congruent to another one, we can ignore it
-            if (d_relevantTerms.find(n) != d_relevantTerms.end()
-                && d_relevantTerms.find(nc) == d_relevantTerms.end())
+            if (rlvSet.find(n) != rlvSet.end()
+                && rlvSet.find(nc) == rlvSet.end())
```
Comment (yoni206) [inline src/theory/strings/base_solver.cpp:219]: Is this clang formatted?

File: src/theory/strings/term_registry.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/strings/term_registry.h:247]: Yes, I tend to use this naming scheme.

The comment mentions when this is computed, with the assumption that you can only call `getRelevantTermSet` during a full effort check.  I can add some further assertions to ensure this method is being used properly.

File: src/theory/strings/base_solver.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/strings/base_solver.cpp:218]: Yes

File: src/theory/strings/term_registry.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/strings/term_registry.h]: I've refactored and improved this, PTAL.

**Review by yoni206 — APPROVED** [review]

**Pull request merged**